### DISCUSSION TITLE
nested npm scripts

### DISCUSSION THREAD

**Discussion by Pomax** [npm/feedback#374] [CLI] [answered]
Back in 2015 when npm-run-all and related tooling had just popped up, and the idea of npm scripts were super new, there was the suggestion to add nested scripts, initially filed and commented on in https://github.com/npm/npm/pull/9970 where the conclusion at the time was that it would probably break past versions of NPM. 

However, the versions of Node and NPM involved at the time are no longer even in support anymore, and so it would be a good time to rethink this: times have changed and the `engines` field exists, so as long as folks add an `"engines": { npm": "..." }` entry to their package.json, older versions of NPM should be able to report that they won't work. As such, I'd like to pitch the idea of nested scripts again. For example, the following set of scripts that rely on `npm-run-all`:

```json
"start": "run-s build:clean && run-p build:js:prod build:common",
"build:clean": "shx rm -rf ./site/frontend && shx mkdir -p ./site/frontend/_js",
"build:common": "run-p build:images build:sass build:contribute",
"build:contribute": "shx cp contribute.json ./site/frontend/",
"build:images": "shx rm -rf ./site/frontend/_images && shx cp -r source/images ./site/frontend/_images",
"build:js:dev": "node esbuild.config.js -- --node-env development",
"build:js:prod": "node esbuild.config.js -- --node-env production",
"build:sass": "run-s build:sass:clean && run-p build:sass:main build:sass:bg && run-s optimize:css",
"build:sass:clean": "shx mkdir -p ./site/frontend/_css",
"build:sass:main": "sass source/sass/main.scss ./site/frontend/_css/main.compiled.css",
"build:sass:bg": "sass source/sass/buyers-guide/bg-main.scss ./site/frontend/_css/buyers-guide.compiled.css",
"optimize:css": "...",
"optimize:js": "...",
```

would be easier to read and maintain as:

```json
"scripts": {
    "start": "run-s build:clean && run-p build:js:prod build:common",
    "build": {
        "clean": "shx rm -rf ./site/frontend && shx mkdir -p ./site/frontend/_js",
        "common": "run-p build:images build:sass:start build:contribute",
        "contribute": "shx cp contribute.json ./site/frontend/",
        "images": "shx rm -rf ./site/frontend/_images && shx cp -r source/images ./site/frontend/_images",
        "js": {
            "dev": "node esbuild.config.js -- --node-env development",
            "prod": "node esbuild.config.js -- --node-env production",
        },
        "sass": {
            "start": "run-s build:sass:clean && run-p build:sass:main build:sass:bg && run-s build:optimize:css",
            "clean": "shx mkdir -p ./site/frontend/_css",
            "main": "sass source/sass/main.scss ./site/frontend/_css/main.compiled.css",
            "bg": "sass source/sass/buyers-guide/bg-main.scss ./site/frontend/_css/buyers-guide.compiled.css",
        },
        "optimize": {
            "css": "...",
            "js": "..."
        }
    }
},
"engines": {
    "npm": "..."
}

```

And no, this would certainly not work for every existing project, but for new projects, and projects that keep up with node/npm releases, having the option to better structure npm scripts would be quite valuable.

Even more so if NPM had syntax that was similar to `npm-run-all`'s `run-s` and `run-p` built in, with dot syntax rather than colon syntax, so that one could say:

```json
"scripts": {
    ...
    "build": {
        ...
        "sass": {
            "start": "npmr --serial build.sass.clean --parallel build.sass.main build.sass.bg --serial optimize.css",
            ...
        }
    }
},
```

Again, locked behind an `engines` entry for the version of NPM that actually supports that new syntax, so that older versions of NPM don't break on this package.json (but instead just go "I can't run this, please upgrade to ...").

**Comment by ljharb** [comment]
Unfortunately, `engines.npm` affects downstream consumers too, not just developers, although scripts only affect developers - so this isn't a viable solution for anyone who wants to retain back compat (ie, to be installed via an older npm version).

**Reply by Pomax** [reply]
But that's fine? This isn't a "solution", it's a new feature that is compatible with the old feature: all older versions of npm still in support will go "I can't run this code", so if you need a project that _has_ to be backward compatible, nothing changes for you and you just keep using flat scripts. The last thing I want is to _replace_ the way scripts work, this is about _enhancing_ it to do more than it does now, while still working in exactly the same way as it does now, if you use the same code as we use now.

Many folks don't need backward compatibility, and as time goes on, that becomes increasingly more true; there are plenty of projects getting uplifted to the latest LTS on the regular, and plenty of new code gets written that has no obligation to be backward compatible with last year's node or npm, so for folks with (initially) that luxury (and as time goes on, that just becomes "everyone") they get a genuine improvement.

Nothing changes for folks who need backward compat, and things change for the better for folks who want to write modern (whatever that is when they sit down to write it) code. And then all the legacy projects automatically catch up over the course of a year or two because of LTS cycling, and everyone wins.

**Comment by MylesBorins** [comment, marked as answer]
hey @Pomax. If you would like to make a proposal for changes to the npm CLI the best next step would be to open an issue over at https://github.com/npm/rfcs

Thanks!

**Reply by Pomax** [reply]
Cheers. I posted here because that's where the npm issue funnel takes me: head on over to https://github.com/npm/cli/issues/new/choose and click "feature request", and it moves you here; might be a good idea to change that to point to the `rfcs` repo?
